test/e2e/kcp_remediations.go

```diff
@@ -0,0 +1,693 @@
+		Namespace: input.namespace,
+		Name:      clusterName,
+	}, input.E2EConfig.GetIntervals(input.specName, "wait-cluster")...)
+
+	return clusterResources
+}
+
+type sendSignalToBootstrappingMachineInput struct {
+	client    client.Client
+	namespace string
+	machine   string
+	signal    string
+}
+
+// sendSignalToBootstrappingMachine sends a signal to a machine stuck during bootstrap.
+func sendSignalToBootstrappingMachine(ctx context.Context, input sendSignalToBootstrappingMachineInput) {
+	log.Logf("Sending bootstrap signal %s to Machine %s", input.signal, input.machine)
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapName,
+			Namespace: input.namespace,
+		},
+	}
+	Expect(input.client.Get(ctx, client.ObjectKeyFromObject(cm), cm)).To(Succeed(), "failed to get mhc-test config map")
+
+	cmWithSignal := cm.DeepCopy()
+	cmWithSignal.Data[configMapDataKey] = input.signal
+	Expect(input.client.Patch(ctx, cmWithSignal, client.MergeFrom(cm))).To(Succeed(), "failed to patch mhc-test config map")
+
+	log.Logf("Waiting for Machine %s to acknowledge signal %s has been received", input.machine, input.signal)
+	Eventually(func() string {
+		_ = input.client.Get(ctx, client.ObjectKeyFromObject(cmWithSignal), cmWithSignal)
+		return cmWithSignal.Data[configMapDataKey]
+	}, "1m", "10s").Should(Equal(fmt.Sprintf("ack-%s", input.signal)), "Failed to get ack signal from machine %s", input.machine)
+
+	machine := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      input.machine,
+			Namespace: input.namespace,
+		},
+	}
+	Expect(input.client.Get(ctx, client.ObjectKeyFromObject(machine), machine)).To(Succeed())
+
+	// Resetting the signal in the config map
+	cmWithSignal.Data[configMapDataKey] = "hold"
+	Expect(input.client.Patch(ctx, cmWithSignal, client.MergeFrom(cm))).To(Succeed(), "failed to patch mhc-test config map")
+}
+
+type waitForMachinesInput struct {
+	lister                          framework.Lister
+	namespace                       string
+	clusterName                     string
+	expectedReplicas                int
+	expectedOldMachines             []string
+	expectedDeletedMachines         []string
+	waitForMachinesIntervals        []interface{}
+	checkMachineListStableIntervals []interface{}
+}
+
+// waitForMachines waits for machines to reach a well known state defined by number of replicas, a list of machines to exists,
+// a list of machines to not exists anymore. The func also check that the state is stable for some time before
+// returning the list of new machines.
+func waitForMachines(ctx context.Context, input waitForMachinesInput) (allMachineNames, newMachineNames []string) {
+	inClustersNamespaceListOption := client.InNamespace(input.namespace)
+	matchClusterListOption := client.MatchingLabels{
+		clusterv1.ClusterNameLabel:         input.clusterName,
+		clusterv1.MachineControlPlaneLabel: "",
+	}
+
+	expectedOldMachines := sets.NewString(input.expectedOldMachines...)
```

Comment:
nit:  let's also change this to the generic set (https://github.com/kubernetes-sigs/cluster-api/pull/8004)